File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob4.tar/Prob4/Prob4/Code/f_serial/poisson.f90
program poisson
    
    use multigrid
    use m_cgsolver
    use matrixconstructor
    
    implicit none
    
    real(8), parameter :: PI = 3.141592653589793
    integer(4) :: nlevel, initial_ngrid

    integer(4), allocatable :: numgrid(:), matrixDOF(:), index_level(:), index2_level(:)
    integer(4) :: sum_matrixDOF, sum2_matrixDOF, index
    integer(4) :: count, ii, jj, kk, i, maxiteration
    real(8)    ::length, tolerance
    real(8), allocatable :: gridsize(:), poissonmatrix(:),rhs(:), solution(:), coordinate(:)
    real(8) :: time_start, time_end
   
    character(LEN=256) :: myfilename

    nlevel = 3
    initial_ngrid = 128
    maxiteration = 20000
    tolerance = 1.0e-10
    length = 1.0

    call CPU_TIME(time_start)

    allocate( numgrid(nlevel) )
    allocate( matrixDOF(nlevel) )
    allocate( gridsize(nlevel) )
    allocate( index_level(nlevel) )
    allocate( index2_level(nlevel) )

    numgrid(nlevel) = initial_ngrid
    gridsize(nlevel) = length/initial_ngrid
    matrixDOF(nlevel) = numgrid(nlevel)*numgrid(nlevel)

    sum_matrixDOF = matrixDOF(nlevel)
    sum2_matrixDOF = matrixDOF(nlevel)*matrixDOF(nlevel)

    do i=nlevel,2,-1
        numgrid(i-1) = numgrid(i)/2
        gridsize(i-1) = length/numgrid(i-1)
        matrixDOF(i-1) = numgrid(i-1)*numgrid(i-1)
        sum_matrixDOF = sum_matrixDOF + matrixDOF(i-1)
        sum2_matrixDOF = sum2_matrixDOF + matrixDOF(i-1)*matrixDOF(i-1)
    enddo
    
    print*,'[Poisson] Geometry and matrix size initialized.'

    allocate( rhs(sum_matrixDOF) )
    allocate( solution(sum_matrixDOF) )
    allocate( coordinate(sum_matrixDOF*2) )
    allocate( poissonmatrix(sum2_matrixDOF) )
    
    index_level(1)=1
    index2_level(1)=1
    do i=2,nlevel
        index_level(i)=index_level(i-1)+matrixDOF(i-1)
        index2_level(i)=index2_level(i-1)+matrixDOF(i-1)*matrixDOF(i-1)
    enddo
    
    do i=1,nlevel
        count = 0
        index = 2*index_level(i)-1
        do jj=1,numgrid(i)
            do ii=1,numgrid(i)
                coordinate(index+2*count) = ii*gridsize(i)-0.5*gridsize(i);
                coordinate(index+2*count+1) = jj*gridsize(i)-0.5*gridsize(i);
               	count=count+1
            enddo 
        enddo
    enddo

    count = 0

    index = 2*index_level(nlevel)-1
    do jj=1,numgrid(nlevel)
        do ii=1,numgrid(nlevel)
            rhs(index_level(nlevel)+count) = sin(coordinate(index+2*count)/length*PI)* &
                   sin(coordinate(index+2*count+1)/length*PI)*gridsize(nlevel)*gridsize(nlevel)
            count=count+1
        enddo 
    enddo
    print*,'[Poisson] Geometry and rhs constructed.'
    
    do i=1,nlevel
        do ii=1,matrixDOF(i)*matrixDOF(i)
            poissonmatrix(index2_level(i)+ii-1) = 0.0
        enddo
    enddo
    
    do i=1,nlevel
        call construct_poissonmatrix(1, 1, numgrid(i), numgrid(i), numgrid(i), poissonmatrix(index2_level(i)))
    enddo

    print*,'[Poisson] Poisson matrix constructed.'
    
    do i=1,matrixDOF(nlevel)
        solution(index_level(nlevel)+i-1) = 1.0
    enddo
    
    print*,'[Poisson] Start solving equations.'
    
    do i=nlevel,2,-1 
        call restriction(rhs(index_level(i)),rhs(index_level(i-1)),numgrid(i-1))
        print 102, i, i-1
        do ii=1,matrixDOF(i-1)
            solution(index_level(i-1)+ii-1) = 1.0
        enddo
    enddo

    call cgsolver(matrixDOF(1), poissonmatrix(index2_level(1)), rhs(index_level(1)), solution(index_level(1)), &
                  maxiteration, tolerance)

    do i=2,nlevel
        call interpolation(solution(index_level(i-1)),solution(index_level(i)),numgrid(i))
        print 103, i-1, i
        call cgsolver(matrixDOF(i), poissonmatrix(index2_level(i)), rhs(index_level(i)), solution(index_level(i)), &
             maxiteration, tolerance)
        print 104, i

    enddo

    print*,'[Poisson] Final solution obtained.'

    call system("rm -rf result")
    call system("mkdir result")

    do i=1,nlevel
        index = 2*index_level(i)-1
        write(myfilename,100) i
        open(40,file=myfilename,status='unknown', form='formatted')
        count=0
        do ii=1,numgrid(i)
            do jj=1,numgrid(i)
                write(40,101) coordinate(index+2*count),coordinate(index+2*count+1), &
                              solution(index_level(i)+count)
                count=count+1
            enddo
        enddo
        
        close(40)
    enddo

    print*,'[Poisson] Final solution printed.'

    deallocate(rhs)
    deallocate(solution)
    deallocate(coordinate)
    deallocate(poissonmatrix)
    
    deallocate( numgrid )
    deallocate( matrixDOF )
    deallocate( gridsize )
    deallocate( index_level )
    deallocate( index2_level )

    print*,'[Poisson] Memory deallocated.'

    call CPU_TIME(time_end)

    write(*,105) time_end-time_start
    
100 format('./result/solution_level',I1,'.dat')
101 format(3(F12.6,x))
102 format(' [Multigrid] Restriction of RHS done (level ',I2.1,' -> level ',I2.1').')
103 format(' [Multigrid] Interpolation of solution done (level ',I2.1,' -> level ',I2.1').')
104 format(' [Poisson] Solution in level ',I2.1,' obtained.')
105 format(' [Poisson] Mission completed in ',F10.6,' (secs).')

end

File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob4.tar/Prob4/Prob4/Code/c_serial/poisson.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "matrixconstructor.h"
#include "cgsolver.h"
#include "multigrid.h"

#define PI 3.141592653589793


int main(int argc, char **argv) 
{
    int nlevel, initial_ngrid;
    int *numgrid, *matrixDOF;
    int count, ii, jj, kk, maxiteration;  
    double length, tolerance, *gridsize;
    double **poissonmatrix, **rhs, **solution, **coordinate;
    clock_t time_start, time_end;
    double elapsed_time;

    char myfilename[256];
    FILE *fp;

    nlevel = 3;
    initial_ngrid = 128;
    maxiteration = 20000;
    tolerance = 1.0e-10;
    length = 1.0;
    length = 1.0;

    time_start = clock();

    numgrid = (int*)malloc(nlevel*sizeof(int));
    matrixDOF = (int*)malloc(nlevel*sizeof(int));
    gridsize = (double*)malloc(nlevel*sizeof(double));

    numgrid[nlevel-1] = initial_ngrid;
    gridsize[nlevel-1] = length/initial_ngrid;
    matrixDOF[nlevel-1] = numgrid[nlevel-1]*numgrid[nlevel-1];

    for(ii=nlevel-1; ii>=1; ii--)
    {
	numgrid[ii-1] = numgrid[ii]/2;
	gridsize[ii-1] = length/numgrid[ii-1];
	matrixDOF[ii-1] = numgrid[ii-1]*numgrid[ii-1];
    }

    printf("[Poisson] Geometry and matrix size initialized.\n");

    poissonmatrix = (double**)malloc(nlevel*sizeof(double*));
    rhs = (double**)malloc(nlevel*sizeof(double*));
    solution = (double**)malloc(nlevel*sizeof(double*));
    coordinate = (double**)malloc(nlevel*sizeof(double*));

    for(ii=0; ii<nlevel; ii++) 
    {
	poissonmatrix[ii] = (double*)malloc(matrixDOF[ii]*matrixDOF[ii]*sizeof(double));
	rhs[ii] = (double*)malloc(matrixDOF[ii]*sizeof(double));
	solution[ii] = (double*)malloc(matrixDOF[ii]*sizeof(double));
	coordinate[ii] = (double*)malloc(matrixDOF[ii]*2*sizeof(double));
    }

    for(ii=0; ii<nlevel; ii++)
    {
	count = 0;

	for(jj=1; jj<=numgrid[ii]; jj++)
	{
	    for(kk=1; kk<=numgrid[ii]; kk++)
	    {
		coordinate[ii][2*count] = jj*gridsize[ii]-0.5*gridsize[ii];
		coordinate[ii][2*count+1] = kk*gridsize[ii]-0.5*gridsize[ii];
		count++;
	    } 
	}
    }

    count = 0;

    for(ii=1; ii<=numgrid[nlevel-1]; ii++)
    {
	for(jj=1; jj<=numgrid[nlevel-1]; jj++)
	{
	    rhs[nlevel-1][count] = sin(coordinate[nlevel-1][2*count]/length*PI)* \
				   sin(coordinate[nlevel-1][2*count+1]/length*PI)*gridsize[nlevel-1]*gridsize[nlevel-1];
	    count++;
	} 
    }

    printf("[Poisson] Geometry and rhs constructed.\n");

    for(ii=0; ii<nlevel; ii++) 
	for(jj=0; jj<matrixDOF[ii]*matrixDOF[ii]; jj++)	
	    poissonmatrix[ii][jj] = 0.0;

    for(ii=0; ii<nlevel; ii++)
	construct_poissonmatrix(1, 1, numgrid[ii], numgrid[ii], numgrid[ii], poissonmatrix[ii]);

//        for(jj=0; jj<matrixDOF[0]*matrixDOF[0]; jj++)
//            printf("poisson=%f\n", poissonmatrix[0][jj]);

    printf("[Poisson] Poisson matrix constructed.\n");

    for(ii=0; ii<matrixDOF[nlevel-1]; ii++)
	solution[nlevel-1][ii] = 1.0;

    printf("[Poisson] Start solving equations.\n");

    for(ii=nlevel-1; ii>0; ii--) 
    {
	restriction(rhs[ii], rhs[ii-1], numgrid[ii-1]);

//        for(int aa=0; aa<numgrid[ii-1]*numgrid[ii-1]; aa++)
//                printf("%f\n", rhs[ii-1][aa]);

	printf("[Multigrid] Restriction of RHS done (level %d -> level %d).\n", ii, ii-1);

	for(jj=0; jj<matrixDOF[ii-1]; jj++)
	    solution[ii-1][jj] = 1.0;
    }

    cgsolver(matrixDOF[0], poissonmatrix[0], rhs[0], solution[0], maxiteration, tolerance);

    for(ii=1; ii<nlevel; ii++)
    {
	interpolation(solution[ii-1], solution[ii], numgrid[ii]);
	printf("[Multigrid] Interpolation of solution done (level %d -> level %d).\n", ii-1, ii);
	cgsolver(matrixDOF[ii], poissonmatrix[ii], rhs[ii], solution[ii], maxiteration, tolerance);
	printf("[Poisson] Solution in level %d obtained.\n", ii);
    }

    printf("[Poisson] Final solution obtained.\n");

    system("rm -rf result\n");
    system("mkdir result\n");

    for(ii=0; ii<nlevel; ii++) 
    {
	sprintf(myfilename,"./result/solution_level%d.dat", ii);
	fp = fopen(myfilename, "wt");

	count = 0;
	for(jj=1; jj<=numgrid[ii]; jj++)
	{
	    for(kk=1; kk<=numgrid[ii]; kk++)
	    {
		fprintf(fp,"%9.5f%10.5f%13.6f\n", coordinate[ii][2*count], coordinate[ii][2*count+1], solution[ii][count]);
		count++;
	    }
	}

	fclose(fp);
    }

    printf("[Poisson] Final solution printed.\n");

    for(ii=0; ii<nlevel; ii++) 
    {
	free(coordinate[ii]);
	free(solution[ii]);
	free(rhs[ii]);
	free(poissonmatrix[ii]);
    }

    free(poissonmatrix);
    free(rhs);
    free(solution);
    free(coordinate);
    free(matrixDOF);
    free(numgrid);
    free(gridsize);

    printf("[Poisson] Memory deallocated.\n");

    time_end = clock();
    elapsed_time = (double)(time_end - time_start)/CLOCKS_PER_SEC;

    printf("[Poisson] Mission completed in %f (secs).\n", elapsed_time);

    return 0;
}


